### PR TITLE
Intrinsics + CUDA [simd]

### PR DESCRIPTION
**Pull request by camierjs**
Allow SIMD with CUDA builds by guarding the intrinsic headers when the `cu` language is used.
Should address issue #1689.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1695](https://github.com/mfem/mfem/pull/1695) | @camierjs | @tzanio | @v-dobrev + @tzanio | 08/13/20 | ⌛due 06/30/21 | ⌛due 07/07/21 | |

### CODE REVIEW

File: linalg/simd/vsx.hpp [hunk 1]
```diff
@@ -12,6 +12,10 @@
 #ifndef MFEM_SIMD_VSX_HPP
 #define MFEM_SIMD_VSX_HPP
 
+#if !(defined(__CUDACC__) || defined(__HIPCC__))
```
Comment (v-dobrev) [inline linalg/simd/vsx.hpp:15]: This means you can never use these intrinsics with CUDA/HIP. When using `nvcc`, is there no host compiler with which these work?

File: linalg/simd/x86.hpp [hunk 2]
```diff
@@ -12,10 +12,14 @@
 #ifndef MFEM_SIMD_X86_HPP
 #define MFEM_SIMD_X86_HPP
 
+#if !(defined(__CUDACC__) || defined(__HIPCC__))
```
Comment (v-dobrev) [inline linalg/simd/x86.hpp:15]: This definitely should be removed -- at least with `nvcc` the x86 intrinsics work fine.

File: miniapps/performance/makefile [hunk 3]
```diff
@@ -82,6 +82,10 @@ endef
 
 ifneq (YES,$(MFEM_USE_CUDA))
   PEDANTIC_FLAG = -pedantic
+else
+	ifeq (YES,$(MFEM_USE_SIMD))
+		MFEM_FLAGS := $(filter-out -x=cu, $(MFEM_FLAGS))
+	endif
```
Comment (v-dobrev) [inline miniapps/performance/makefile:88]: This seems redundant. Was there a specific reason to add it?

Comment (v-dobrev) [reply miniapps/performance/makefile:88]: I think I understand why you had to add this: without `-x=cu` when a `.cpp` file is compiled, `nvcc` does not define `__CUDACC__`, so you get the intrinsics definitions.

This is a convoluted way of enabling the intrinsics. Isn't there a more streamlined way of achieving this?

Comment (camierjs) [reply miniapps/performance/makefile:88]: Right - I agree this solution does not allow to have both `MFEM_FORALL` **and** `intrinsics` in the same file, which we actually dont have in `MFEM`.
The issue comes with `nvcc` including these headers, but won't ever use them.

**Comment by tzanio** [comment]
Postponed...

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Comment by tzanio** [comment]
Moved to WIP until we have a better solution (@v-dobrev)

The fix in linalg/simd/vsx128.hpp will be brought in another PR (@camierjs)

**Comment by camierjs** [comment]
Closing as the fix in linalg/simd/vsx128.hpp has been brought into master.

**Pull request closed without merging**